docs/design/query-caching.md

```diff
@@ -0,0 +1,224 @@
+
+### Objectives
+
+- Cache responses to read operations
+    - REST: GET requests REST
+    - GraphQL: Queries
+- Make cache configuration simple.
+    - Limit developer exposure to implementation details of the cache and only surface essential properties in the runtime configuration.
+    - Runtime config properties remain consistent and optional for backwards and future compatibility.
+    - Replacing cache implementation (e.g. libraryA to libraryB) does not observably affect developers and their configuration.
+
+### Non-Objectives
+
+-	Follow-up reads on mutations (REST/GraphQL) are not cached.
+-	Distributed caching for syncing cache entries across servers for multiple instances of DAB.
+
+## Design
+
+The caching service should be database and endpoint agnostic. DAB will instantiate a singleton caching service that can be used within dependent services. In order to reduce coupling between DAB and a particular caching service implementation, the most generic caching options are publicly exposed in the runtime configuration: cache enablement and cache entry ttl. Additional options specific to the caching service will be explicitly set internally by DAB.
+
+Intentionally surfacing limited configuration options to the developer enables us to __make the caching service function uniformaly and independent of local hosting and SWA hosting configuration.__ This is because the developer exposed configuration properties are applied per-request and unnecessary for configuring the cache service during application startup.
+
+To reduce the sprawl of cache implementation across the code base, we will introduce the caching layer within __a common step of query execution for both REST and GraphQL: the query engine.__ The query engine's responsibility is to build a database query and dispatch the query to the appropriate query executor aligned to the backend database type. 
+
+By keeping the caching layer out of the query executor layer, we eliminate dependencies between database specific query execution logic and caching. Additionally, we prevent adding even tighter coupling between the runtime configuration and query executor because the query executor will not need to reference the runtime config to determine whether caching a request is necessary.
+
+### Cache entry
+
+#### Key
+
+`dataSourceName:queryText:queryParameters`
+
+The following example illustrates a real value created using the key schema listed above (new lines added for readability):
+
+```text
+7a07f92a-1aa2-4e2a-81d6-b9af0a25bbb6:
+
+SELECT TOP 1 [dbo_books].[id] AS [id], [dbo_books].[title] AS [title], [dbo_books].[publisher_id] AS [publisher_id] FROM [dbo].[books] AS [dbo_books] WHERE [dbo_books].[id] = @param0 ORDER BY [dbo_books].[id] ASC FOR JSON PATH, INCLUDE_NULL_VALUES,WITHOUT_ARRAY_WRAPPER:
+
+{"@param0":{"Value":1,"DbType":11},"@param1":{"Value":"id","DbType":null},"@param2":{"Value":"title","DbType":null},"@param3":{"Value":"publisher_id","DbType":null}}
```

Comment:
Order here will matter, will this meta data always be ordered with Param0 first, etc? If not, it may need to be sorted to ensure you get consistent keys.

Reply:
Even if params are in order, there may be possibility that their values are not 1:1 across generated queries when users change up the order of fields in the request: 
- REST: change in order of composite PK, or order of query string params
- GraphQL: change in order of fields in selection set.

I observed that other caching services caveat cache performance is dependent on ordering of such values I mention above. We can document that too. We'd have to do perf tests to determine whether sorting keys and values adds overhead that defeats the purpose of a cache.

Reply:
that's a good point about other orderings being important too. I would say one thing though, I would have less sympathy for someone suffering from cache misses if theyre swapping around the ordering in their request on their own versus if the ordering is coming out differently just because of how some data structure is being iterated under the hood. not sure how easy it is to separate those, but we could certainly condition the cache performance on the user not messing with the order without me losing any sleep :)